Add new gemini models and update the safety settings

sirius422 opened the pull request:
# Add New Gemini Models (Feb 2025)

Added the new Gemini models that Google released on Feb 5th:
- `gemini-2.0-flash`
- `gemini-2.0-flash-001` 
- `gemini-2.0-flash-lite-preview`
- `gemini-2.0-flash-lite-preview-02-05`
- `gemini-2.0-pro-exp`
- `gemini-2.0-pro-exp-02-05`

Also fixed my previous oversight - updated the `visionSupportedModels` in `public/scripts/openai.js` this time 😅

## Safety Settings Update

After some testing, I found that some older models do not support setting `threshold` to `OFF` and only allow `BLOCK_NONE`:
- `gemini-1.5-pro-001`
- `gemini-1.5-flash-001`
- `gemini-1.0-pro-001`

However, `gemini-2.0-flash` models can have all thresholds set to `OFF`:
- `gemini-2.0-flash` 
- `gemini-2.0-flash-001`
- `gemini-2.0-flash-exp`  

For the other models, setting all thresholds to `OFF` will result in an error:
```json
{
  "error": {
    "code": 400,
    "message": "HARM_CATEGORY_CIVIC_INTEGRITY threshold cannot be 5",
    "status": "INVALID_ARGUMENT"
  }
}
```

Therefore, I updated the default `GEMINI_SAFETY` settings to:
```javascript
{
    category: 'HARM_CATEGORY_HARASSMENT',
    threshold: 'OFF',
},
{
    category: 'HARM_CATEGORY_HATE_SPEECH',
    threshold: 'OFF',
},
{
    category: 'HARM_CATEGORY_SEXUALLY_EXPLICIT',
    threshold: 'OFF',
},
{
    category: 'HARM_CATEGORY_DANGEROUS_CONTENT',
    threshold: 'OFF',
},
{
    category: 'HARM_CATEGORY_CIVIC_INTEGRITY',
    threshold: 'BLOCK_NONE',
},
```

<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

Commit:
feat: update Gemini models

- Add new Gemini models (2025/02/05)

Commit:
feat: update Gemini safety settings

- Set threshold to OFF for most models, except for HARM_CATEGORY_CIVIC_INTEGRITY.
- Handle specific cases for a few models.

Wolfsblvt commented:
Whelp, I am sorry. I didn't see the PR before I quickly added in the models.
This is a lot better and more thorough. But with the backend changes, might require some actual testing.

Gonna look at this tomorrow, if Cohee doesn't.

Wolfsblvt committed:
Merge branch 'staging' into pr/3434

Wolfsblvt reviewed (approved):
Okay, gave this a thorough-ish test.
All models seem to work, tested a few vision models for image captioning, and I did not get any errors on any of threshold switches you added.

Trust that you thoroughly checked that.

As multiple people were asking for this, I'll go ahead and merge.

cloak1505 commented:
Cleanup required.

gemini-1.5-flash-8b-exp-0827, gemini-1.5-flash-8b-exp-0924, gemini-pro, gemini-1.0-pro (other 1.5 models still okay):
```
{"code":400,"message":"You cannot set safety setting as OFF for this model.","status":"INVALID_ARGUMENT"}
```

gemini-exp-1114, gemini-exp-1121, gemini-1.5-pro-exp-0801, gemini-1.5-pro-exp-0827 (gemini-exp-1206 still okay):
```
{"code":404,"message":"models/gemini-exp-1114 is not found for API version v1beta, or is not supported for generateContent. Call ListModels to see the list of available models and their supported methods.","status":"NOT_FOUND"}
```

gemini-ultra, gemini-1.0-ultra-latest:
```
{"code":404,"message":"models/gemini-ultra is not found for API version v1beta, or is not supported for generateContent. Call ListModels to see the list of available models and their supported methods.","status":"NOT_FOUND"}
```

Edit: Gemini 1.0 Pro will be removed on Feb. 15.

sirius422 commented:
> Cleanup required.
> 
> gemini-1.5-flash-8b-exp-0827, gemini-1.5-flash-8b-exp-0924, gemini-pro, gemini-1.0-pro (other 1.5 models still okay):
> 
> ```
> {"code":400,"message":"You cannot set safety setting as OFF for this model.","status":"INVALID_ARGUMENT"}
> ```
> 
> gemini-exp-1114, gemini-exp-1121, gemini-1.5-pro-exp-0801, gemini-1.5-pro-exp-0827 (gemini-exp-1206 still okay):
> 
> ```
> {"code":404,"message":"models/gemini-exp-1114 is not found for API version v1beta, or is not supported for generateContent. Call ListModels to see the list of available models and their supported methods.","status":"NOT_FOUND"}
> ```
> 
> gemini-ultra, gemini-1.0-ultra-latest:
> 
> ```
> {"code":404,"message":"models/gemini-ultra is not found for API version v1beta, or is not supported for generateContent. Call ListModels to see the list of available models and their supported methods.","status":"NOT_FOUND"}
> ```
> 
> Edit: Gemini 1.0 Pro will be removed on Feb. 15.

@cloak1505 Thanks for your testing. I wrote a script to test different models' responses, but I indeed missed testing these ones:
- gemini-1.5-flash-8b-exp-0827
- gemini-1.5-flash-8b-exp-0924
- gemini-pro
- gemini-1.0-pro

I didn't test the two experimental gemini-1.5-flash-8b models because from the `https://generativelanguage.googleapis.com/v1beta/models` response JSON, I knew they were replaced by `gemini-1.5-flash-8b-001`:
```json
[
    {
        "name": "models/gemini-1.5-flash-8b-exp-0827",
        "version": "001",
        "displayName": "Gemini 1.5 Flash 8B Experimental 0827",
        "description": "Experimental release (August 27th, 2024) of Gemini 1.5 Flash-8B, our smallest and most cost effective Flash model. Replaced by Gemini-1.5-flash-8b-001 (stable).",
        "inputTokenLimit": 1000000,
        "outputTokenLimit": 8192,
        "supportedGenerationMethods": [
            "generateContent",
            "countTokens"
        ],
        "temperature": 1,
        "topP": 0.95,
        "topK": 40,
        "maxTemperature": 2
    },
    {
        "name": "models/gemini-1.5-flash-8b-exp-0924",
        "version": "001",
        "displayName": "Gemini 1.5 Flash 8B Experimental 0924",
        "description": "Experimental release (September 24th, 2024) of Gemini 1.5 Flash-8B, our smallest and most cost effective Flash model. Replaced by Gemini-1.5-flash-8b-001 (stable).",
        "inputTokenLimit": 1000000,
        "outputTokenLimit": 8192,
        "supportedGenerationMethods": [
            "generateContent",
            "countTokens"
        ],
        "temperature": 1,
        "topP": 0.95,
        "topK": 40,
        "maxTemperature": 2
    }
]
```
Now it appears they do have differences in accepted request parameters, so these two should be included in the validation logic.

As for gemini-pro and gemini-1.0-pro, I overlooked that gemini-pro is an alias for gemini-1.0-pro.

------

`gemini-exp-1114`, `gemini-exp-1121`, `gemini-1.5-pro-exp-0801`, and `gemini-1.5-pro-exp-0827` no longer exist in models.json, so calls to them will fail directly, making safety-related validation unnecessary. It's quite strange - these two experimental models gemini-1.5-flash-8b models (0827/0924) still have working endpoints, even though they're older than several experimental models that have already been removed from models.json.

Regarding `gemini-ultra` and `gemini-1.0-ultra-latest`, they previously required contacting Google for access permissions. I'm unsure if Google still provides endpoints for these models and couldn't test them. According to Reddit reports, Ultra might have been removed as early as May last year.

------

I'll open a new Pull Request to fix this bug. As for gemini-ultra and outdated gemini experimental models, I noticed @Wolfsblvt  previously suggested cleaning up the model list in #3437. Though perhaps a better approach would be to query `v1beta/models` to get all currently available models.
